diagnostics/reporter.rs

```diff
@@ -0,0 +1,159 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    cell::RefCell,
+    collections::HashMap,
+    fs,
+    hash::{DefaultHasher, Hash, Hasher},
+    io::{self, Write},
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Mutex,
+    },
+    thread,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use chrono::{DateTime, Timelike, Utc};
+use concurrency::IntervalRunner;
+use logger::{debug, trace};
+use reqwest::{
+    blocking::{Client, Response},
+    header::{HeaderValue, CONNECTION, CONTENT_TYPE},
+};
+use resource::constants::{
+    common::SECONDS_IN_MINUTE,
+    diagnostics::{DISABLED_REPORTING_FILE_NAME, REPORT_INTERVAL, REPORT_ONCE_DELAY},
+};
+
+use crate::{hash_string_consistently, Diagnostics};
+
+#[derive(Debug)]
+pub struct Reporter {
```

Comment:
I'm not yet sure if I want to use this class purely for the diagnostics service or for both the diagnostics service and PostHog. Tight coupling can become painful in the future if we want to configure them separately. I may refactor this class in the next PR with the PostHog introduction.